tests/test_json_patch.c

```diff
@@ -0,0 +1,119 @@
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "json.h"
+
+#ifndef PATH_MAX
+#define PATH_MAX 256
+#endif
+
+void test_json_patch_op(struct json_object *jo)
+{
+	const char *comment = json_object_get_string(json_object_object_get(jo, "comment"));
+	struct json_object *doc = json_object_object_get(jo, "doc");
+	struct json_object *patch = json_object_object_get(jo, "patch");
+	struct json_object *expected = json_object_object_get(jo, "expected");
+	struct json_object *error = json_object_object_get(jo, "error");
+	int disabled_test = json_object_get_boolean(json_object_object_get(jo, "disabled_in_json_c"));
+	const char *error_s = json_object_get_string(error);
+	struct json_object *res = NULL;
+	int ret;
+
+	printf("Testing '%s', doc '%s' patch '%s' : ",
+		comment ? comment : error_s,
+		json_object_get_string(doc),
+		json_object_get_string(patch));
+	if (disabled_test) {
+		printf("SKIPPING - disabled in the test spec\n");
+		return;
+	}
+	if (!error && !expected) {
+		printf("SKIPPING - no expected or error conditions in test\n");
+		return;
+	}
+	fflush(stdout);
+	if (error) {
+		assert(-1 == json_patch_apply(doc, patch, &res));
+		assert(res == NULL);
+	} else {
+		ret = json_patch_apply(doc, patch, &res);
+		if (ret) {
+			fprintf(stderr, "json_patch_apply() returned '%d'\n", ret);
+			fprintf(stderr, "Expected: %s\n", json_object_get_string(expected));
+			fprintf(stderr, "Got: %s\n", json_object_get_string(res));
+			fflush(stderr);
+			assert(0);
+		}
+		assert(res != NULL);
+		ret = json_object_equal(expected, res);
+		if (ret == 0) {
+			fprintf(stderr, "json_object_equal() returned '%d'\n", ret);
+			fprintf(stderr, "Expected: %s\n", json_object_get_string(expected));
+			fprintf(stderr, "Got: %s\n", json_object_get_string(res));
+			fflush(stderr);
+			assert(0);
+		}
+		json_object_put(res);
+		res = NULL;
+	}
+
+	printf("OK\n");
+}
+
+void test_json_patch_using_file(const char *testdir, const char *filename)
+{
+	char full_filename[PATH_MAX];
+	(void)snprintf(full_filename, sizeof(full_filename), "%s/%s", testdir, filename);
```

Comment:
You need a #include "snprintf_compat.h" at the top of this file to use snprintf.  (otherwise Windows builds break)

Reply:
done in https://github.com/hawicz/json-c/tree/json_patch